demos/TicTacToe/Classes/MainMenuScene.cpp

```diff
@@ -22,35 +91,316 @@ bool MainMenuScene::init() {
-  // scene with a TicTacToe scene.
-  create_button_touch_listener->onTouchBegan = [](Touch* touch,
-                                                  Event* event) -> bool {
-    auto bounds = event->getCurrentTarget()->getBoundingBox();
-    auto point = touch->getLocation();
-    // Replaces the scene with a new TicTacToe scene if the touched point is
-    // within the bounds of the button.
+  auto login_button_touch_listener = EventListenerTouchOneByOne::create();
+
+  // Transition from kAuthState to kWaitingLoginState on button press and set
+  // user_result_ to SignInWithEmailAndPassword future result.
+  login_button_touch_listener->onTouchBegan = [this](Touch* touch,
+                                                     Event* event) -> bool {
+    // Returns if the layer is not in the auth state or is switching states.
+    if (previous_state_ != current_state_ || current_state_ != kAuthState) {
+      return true;
+    }
+    const auto bounds = event->getCurrentTarget()->getBoundingBox();
+    const auto point = touch->getLocation();
     if (bounds.containsPoint(point)) {
-      Director::getInstance()->replaceScene(
-          TicTacToe::createScene(std::string()));
+      // Check the string for a valid email pattern.
+      if (!std::regex_match(email_text_field_->getString(), email_pattern)) {
+        invalid_login_label_->setString("invalid email address");
+      } else if (password_text_field_->getString().length() < 8) {
+        invalid_login_label_->setString(
+            "password must be at least 8 characters long");
+      } else {
+        user_result_ = auth_->SignInWithEmailAndPassword(
+            email_text_field_->getString().c_str(),
+            password_text_field_->getString().c_str());
+        current_state_ = kWaitingLoginState;
+      }
     }
+    return true;
+  };
+
+  // Attach the touch listener to the login button.
+  Director::getInstance()
+      ->getEventDispatcher()
+      ->addEventListenerWithSceneGraphPriority(login_button_touch_listener,
+                                               login_button);
+  auth_background_->addChild(login_button, 1);
 
+  // Create the sign_up button and give it a position, anchor point and
+  // touch_listener.
+  auto sign_up_button = Sprite::create(kSignUpButtonImage);
+  sign_up_button->setPosition(310, 120);
+  sign_up_button->setAnchorPoint(Vec2(0, 0));
+  sign_up_button->setContentSize(Size(200, 75));
+
+  // Create a button listener to handle the touch event.
+  auto sign_up_button_touch_listener = EventListenerTouchOneByOne::create();
+
+  // Transition from kAuthState to kWaitingSignUpState on button press and set
+  // user_result_ to CreateUserWithEmailAndPassword future result.
+  sign_up_button_touch_listener->onTouchBegan = [this](Touch* touch,
+                                                       Event* event) -> bool {
+    // Returns if the layer is not in the auth state or is switching states.
```

Comment:
I would move these "returns if the layer" comments to inside the scoping block.  And I would say something as to why true is return.  Returning is one thing, returning true means something else.